ddtrace/internal/products.py

```diff
@@ -0,0 +1,178 @@
+            name = product_plugin.name
+            log.debug("Discovered product plugin '%s'", name)
+
+            # Load the product protocol object
+            try:
+                product: Product = product_plugin.load()
+            except Exception:
+                log.exception("Failed to load product plugin '%s'", name)
+                continue
+
+            log.debug("Product plugin '%s' loaded successfully", name)
+
+            self.__products__[name] = product
+
+    def _sort_products(self) -> t.List[t.Tuple[str, Product]]:
+        # Data structures for topological sorting
+        q: t.Deque[str] = deque()  # Queue of products with no dependencies
+        g = defaultdict(list)  # Graph of dependencies
+        f = {}  # Remaining dependencies for each product
+
+        for name, product in self.__products__.items():
+            product_requires = getattr(product, "requires", [])
+            if not product_requires:
+                q.append(name)
+            else:
+                f[name] = list(product_requires)
+                for r in product_requires:
+                    g[r].append(name)
+
+        # Determine the product (topological) ordering
+        ordering = []
+        while q:
+            n = q.popleft()
+            ordering.append(n)
+            for p in g[n]:
+                f[p].remove(n)
+                if not f[p]:
+                    q.append(p)
+                    del f[p]
+
+        if f:
+            log.error(
+                "Circular dependencies among products detected. These products won't be enabled: %s.", list(f.keys())
+            )
+
+        return [(name, self.__products__[name]) for name in ordering if name not in f]
+
+    @property
+    def products(self) -> t.List[t.Tuple[str, Product]]:
+        if self._products is None:
+            self._products = self._sort_products()
+        return self._products
+
+    def start_products(self) -> None:
+        for name, product in self.products:
+            try:
+                product.start()
+                log.debug("Started product '%s'", name)
+            except Exception:
+                log.exception("Failed to start product '%s'", name)
+
+    def restart_products(self, join: bool = False) -> None:
+        for name, product in self.products:
+            try:
+                product.restart(join=join)
+                log.debug("Restarted product '%s'", name)
+            except Exception:
+                log.exception("Failed to restart product '%s'", name)
+
+    def stop_products(self, join: bool = False) -> None:
+        for name, product in reversed(self.products):
+            try:
+                product.stop(join=join)
+                log.debug("Stopped product '%s'", name)
+            except Exception:
+                log.exception("Failed to stop product '%s'", name)
+
+    def exit_products(self, join: bool = False) -> None:
+        for name, product in reversed(self.products):
```

Comment:
So we need to do this and `stop_products` in reverse because of the topological sorting?

Reply:
Correct